pkg/client/query/cache/memory.go

```diff
@@ -0,0 +1,304 @@
+package cache
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pokt-network/poktroll/pkg/client"
+)
+
+var (
+	_ client.QueryCache[any]           = (*InMemoryCache[any])(nil)
+	_ client.HistoricalQueryCache[any] = (*InMemoryCache[any])(nil)
+
+	DefaultQueryCacheConfig = queryCacheConfig{
+		EvictionPolicy: FirstInFirstOut,
+		// TODO_MAINNET(@bryanchriswhite): Consider how we can "guarantee" good
+		// alignment between the TTL and the block production rate.
+		TTL: time.Minute,
+	}
+)
+
+// InMemoryCache provides a concurrency-safe in-memory cache implementation with
+// optional historical value support.
+type InMemoryCache[T any] struct {
+	config       queryCacheConfig
+	latestHeight atomic.Int64
+
+	itemsMu sync.RWMutex
+	// items type depends on historical mode:
+	// | historical mode |  type                          |
+	// | --------------- | ------------------------------ |
+	// | false           | map[string]cacheItem[T]        |
+	// | true            | map[string]cacheItemHistory[T] |
+	items map[string]any
```

Comment:
Why items instead of values?

Reply:
Will #PUC.

Reply:
I restructured things a bit and did some renaming:

```go
// inMemoryCache provides a concurrency-safe in-memory cache implementation with
// optional historical value support.
type inMemoryCache[T any] struct {
	config       queryCacheConfig
	latestHeight atomic.Int64

	// valuesMu is used to protect values AND valueHistories from concurrent access.
	valuesMu sync.RWMutex
	// values holds the cached values in non-historical mode.
	values map[string]cacheValue[T]
	// valueHistories holds the cached historical values in historical mode.
	valueHistories map[string]cacheValueHistory[T]
}

// cacheValue wraps cached values with a timestamp for later comparison against
// the configured TTL.
type cacheValue[T any] struct {
	value     T
	timestamp time.Time
}

// cacheValueHistory stores cachedItems by height and maintains a sorted list of
// heights for which cached items exist. This list is sorted in descending order
// to improve performance characteristics by positively correlating index with age.
type cacheValueHistory[T any] struct {
	// sortedDescHeights is a list of the heights for which values are cached.
	// It is sorted in descending order.
	sortedDescHeights []int64
	heightMap         map[int64]cacheValue[T]
}
```

Reply:
This is great, ty!